composeApp/src/commonMain/kotlin/org/ooni/probe/ui/descriptor/AddDescriptorViewModel.kt

```diff
@@ -0,0 +1,117 @@
+package org.ooni.probe.ui.descriptor
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.launch
+import org.ooni.engine.Engine
+import org.ooni.engine.models.Result
+import org.ooni.probe.data.models.InstalledTestDescriptorModel
+import org.ooni.probe.data.models.NetTest
+import org.ooni.probe.ui.shared.SelectableItem
+
+class AddDescriptorViewModel(
+    onCancel: () -> Unit,
+    onError: () -> Unit,
+    fetchDescriptor: suspend () -> Result<InstalledTestDescriptorModel?, Engine.MkException>,
+    saveTestDescriptors: suspend (List<Pair<InstalledTestDescriptorModel, List<NetTest>>>) -> Unit,
+) : ViewModel() {
+    private val events = MutableSharedFlow<Event>(extraBufferCapacity = 1)
+
+    private val _state = MutableStateFlow(State())
+    val state = _state.asStateFlow()
+
+    init {
+        viewModelScope.launch {
+            fetchDescriptor().map { descriptor ->
+                descriptor?.let {
+                    _state.value = State(
+                        descriptor = it,
+                        selectableItems = it.netTests?.map { nettest ->
+                            SelectableItem(
+                                item = nettest,
+                                isSelected = true,
+                            )
+                        } ?: emptyList(),
+                    )
+                }
+            }.onFailure {
+                onError()
```

Comment:
You can also tell the Screen to display the error, and use the onBack here instead. Snackbars are tied to the Activity, so they survive if you switch screens.

Reply:
`onError` shows an error message before going back